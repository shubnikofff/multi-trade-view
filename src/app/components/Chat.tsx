import React from 'react';

interface ChatProps {
}

function Chat() {
    return (
        <div>Chat</div>
    );
}

export default React.memo<ChatProps>(Chat);
