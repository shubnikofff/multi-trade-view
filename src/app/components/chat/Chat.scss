.chat {
  height: 100%;
  overflow-y: auto;
  display: flex;
  flex-direction: column;

  &__header {
    padding: 1rem;
    border-bottom: 1px solid lightgray;
  }

  &__body {
    flex: 1;
    height: 100%;
    overflow-y: auto;
  }

  &__message {
    padding: 1rem;
    border-bottom: 1px solid lightgray;
  }

  &__footer {
    padding: 1rem;
    border-top: 1px solid lightgray;
  }
}
