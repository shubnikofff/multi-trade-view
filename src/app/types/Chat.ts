import { UserRoleEnum } from '@type/User';

export type Message = {
    sendTime: Date;
    text: string;
    sender: UserRoleEnum;
}

export type Chat = {
    id: number;
    hasUnreadMessages: boolean;
    messages: Message[];
}
