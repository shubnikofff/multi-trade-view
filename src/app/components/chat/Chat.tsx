import React, { useRef } from 'react';
import classNames from 'classnames';

import { useChat } from './useChat';

import Avatar from '../avatar';
import NotAvailable from '../not-available';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Formik, Form, Field, FormikHelpers } from 'formik';
import { Link } from 'react-router-dom';

import { faTrashAlt, faChevronLeft, faChevronRight } from '@fortawesome/free-solid-svg-icons'
import { PATH_DASHBOARD, PATH_ROOT } from '../../paths';

import { UserRole } from '../../types/user';

import './Chat.scss';

interface ChatProps {
    smallScreen?: boolean,
}

interface FormValues {
    message: string;
}

function validate({ message }: FormValues) {
    let errors: Partial<FormValues> = {};

    if (!message.trim()) {
        errors.message = 'Required';
    }

    return errors;
}

function Chat({ smallScreen }: ChatProps) {
    const {
        auth,
        chat,
        removeTrade,
        sellerAvatarUrl,
        sendMessage,
        trade,
    } = useChat();

    const chatBodyBottom = useRef<HTMLDivElement>(null);

    if (!trade) {
        return (
            <NotAvailable>
                No such trade
            </NotAvailable>
        );
    }

    const handleSubmit = ({ message }: FormValues, { resetForm }: FormikHelpers<FormValues>) => {
        sendMessage(message);
        chatBodyBottom?.current?.scrollIntoView({ behavior: 'smooth' });
        resetForm();
    }

    return (
        <div className="chat">
            <div className="chat__header">
                <div>
                    <div className="chat__header__left-area">
                        {smallScreen && <div className="chat__header__left-area_back-link">
							<Link to={`${PATH_ROOT}`}>
								<FontAwesomeIcon icon={faChevronLeft} color="black" />
							</Link>
						</div>}
                        <div>
                            <button
                                onClick={removeTrade}
                                disabled={auth === UserRole.Buyer}
                            >
                                <FontAwesomeIcon icon={faTrashAlt} color="white" />
                            </button>
                        </div>
                    </div>
                </div>
                <div className="chat__header__trade-info">
                    <div className="chat__header__trade-info__payment-method">
                        <b>{trade.paymentMethod}</b>
                    </div>
                    <div className="chat__header__trade-info__buyer">
                        <span className="text-secondary">{trade.buyer.name} </span>
                        <span className="text-primary">+{trade.buyer.positiveReputation}</span>
                        <span className="text-secondary"> / </span>
                        <span className="text-danger">-{trade.buyer.negativeReputation}</span>
                    </div>
                </div>
                <div>
                    {smallScreen && <Link to={`${PATH_DASHBOARD}/${trade.id}`}>
						<FontAwesomeIcon icon={faChevronRight} color="black" />
					</Link>}
                </div>
            </div>
            <div className="chat__body">
                {chat && chat.messages.map((message, index) => (
                    <div
                        className={classNames('chat__message', {
                            'chat__message_reversed': message.sender !== auth,
                        })}
                        key={index}
                    >
                        <div className="chat__message__avatar">
                            <Avatar url={message.sender === UserRole.Seller
                                ? sellerAvatarUrl
                                : trade?.buyer.avatarUrl
                            } />
                        </div>
                        <div>
                            <div className={classNames('chat__message__text', message.sender === auth
                                ? 'chat__message__text_outgoing'
                                : 'chat__message__text_incoming'
                            )}>
                                {message.text}
                            </div>
                            <div className={classNames('chat__message__date', {
                                'chat__message__date_align-right': message.sender !== auth,
                            })}>
                                {message.sendTime.toLocaleString()}
                            </div>
                        </div>
                    </div>
                ))}
                <div ref={chatBodyBottom} />
            </div>
            <Formik
                initialValues={{ message: '' }}
                onSubmit={handleSubmit}
                validate={validate}
            >
                <Form>
                    <div className="chat__footer">
                        <div className="chat__footer__left-area">
                            <Field
                                name="message"
                                placeholder="Type your message..."
                            />
                        </div>
                        <div className="chat__footer__right-area">
                            <button type="submit">
                                Send
                            </button>
                        </div>
                    </div>
                </Form>
            </Formik>
        </div>
    );
}

export default React.memo<ChatProps>(Chat);
