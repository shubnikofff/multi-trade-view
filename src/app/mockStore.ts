import configureMockStore from 'redux-mock-store'
import { UserRole } from './types/user';

const mockStore = configureMockStore([]);

const store = mockStore({
    auth: UserRole.Seller,
    trades: {
        ids: [1, 2],
        entities: {
            1: {
                id: 1,
                chatId: 1,
                paymentMethod: 'Amazon Gift Card',
                paid: false,
                buyerId: 1,
                amount: 100,
                hash: 'abc42',
                started: new Date(2020, 8),
            },
            2: {
                id: 2,
                chatId: 2,
                paymentMethod: 'iTunes Gift Card',
                paid: true,
                buyerId: 1,
                amount: 150,
                hash: 'cd23e',
                started: new Date(2020, 8),
            }
        }
    },
    users: {
        ids: [1],
        entities: {
            1: {
                id: 1,
                avatarUrl: 'url',
                name: 'username',
                positiveReputation: 25,
                negativeReputation: 12,
                ofTrades: 4,
            }
        }
    },
    chats: {
        ids: [1, 2],
        entities: {
            1: {
                id: 1,
                hasUnreadMessages: false,
                messages: [
                    {
                        message: 'Message_1',
                        sender: UserRole.Buyer,
                        sendTime: new Date(2020, 8),
                    },
                    {
                        message: 'Message_2',
                        sender: UserRole.Seller,
                        sendTime: new Date(2020, 8),
                    }
                ],
            },
            2: {
                id: 2,
                hasUnreadMessages: false,
                messages: [
                    {
                        message: 'Message_1',
                        sender: UserRole.Buyer,
                        sendTime: new Date(2020, 8),
                    },
                    {
                        message: 'Message_2',
                        sender: UserRole.Seller,
                        sendTime: new Date(2020, 8),
                    }
                ],
            },
        },
    },
    rate: 10000,
});

export { store }
