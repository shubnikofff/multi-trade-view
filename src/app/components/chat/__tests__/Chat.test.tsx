import React from 'react';

import { mount } from 'enzyme';
import { createMemoryHistory } from 'history'

import { Provider } from 'react-redux';
import { Router, Route } from 'react-router-dom'
import Chat from '../Chat';

import { store } from '../../../mockStore';
import { PATH_CHAT } from '../../../constants';

describe('Component chat', () => {
    const history = createMemoryHistory({
        initialEntries: [`${PATH_CHAT}/1`],
        keyLength: 0,
    });

    const wrapper = mount(
        <Provider store={store}>
            <Router history={history}>
                <Route exact path={`${PATH_CHAT}/:tradeId`}>
                    <Chat />
                </Route>
            </Router>
        </Provider>
    );

    afterEach(() => {
        store.clearActions();
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should dispatch actions removeTrade and removeChat', () => {
        const expectedActions = [
            {
                payload: 1,
                type: "trades/removeTrade",
            },
            {
                payload: 1,
                type: "chats/removeChat",
            },
        ];

        wrapper.find('.chat__header__left-area button').simulate('click');
        expect(store.getActions()).toEqual(expect.arrayContaining(expectedActions));
    });
});
