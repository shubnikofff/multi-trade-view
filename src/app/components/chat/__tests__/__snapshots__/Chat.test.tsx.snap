// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component chat should match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/chat/1",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/chat/1",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Route
      exact={true}
      path="/chat/:tradeId"
    >
      <Memo(Chat)>
        <div
          className="chat"
        >
          <div
            className="chat__header"
          >
            <div>
              <div
                className="chat__header__left-area"
              >
                <div>
                  <button
                    disabled={false}
                    onClick={[Function]}
                  >
                    <FontAwesomeIcon
                      border={false}
                      className=""
                      color="white"
                      fixedWidth={false}
                      flip={null}
                      icon={
                        Object {
                          "icon": Array [
                            448,
                            512,
                            Array [],
                            "f2ed",
                            "M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z",
                          ],
                          "iconName": "trash-alt",
                          "prefix": "fas",
                        }
                      }
                      inverse={false}
                      listItem={false}
                      mask={null}
                      pull={null}
                      pulse={false}
                      rotation={null}
                      size={null}
                      spin={false}
                      swapOpacity={false}
                      symbol={false}
                      title=""
                      transform={null}
                    >
                      <svg
                        aria-hidden="true"
                        className="svg-inline--fa fa-trash-alt fa-w-14 "
                        color="white"
                        data-icon="trash-alt"
                        data-prefix="fas"
                        focusable="false"
                        role="img"
                        style={Object {}}
                        viewBox="0 0 448 512"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                          fill="currentColor"
                          style={Object {}}
                        />
                      </svg>
                    </FontAwesomeIcon>
                  </button>
                </div>
              </div>
            </div>
            <div
              className="chat__header__trade-info"
            >
              <div
                className="chat__header__trade-info__payment-method"
              >
                <b>
                  Amazon Gift Card
                </b>
              </div>
              <div
                className="chat__header__trade-info__buyer"
              >
                <span
                  className="text-secondary"
                >
                  username
                   
                </span>
                <span
                  className="text-primary"
                >
                  +
                  25
                </span>
                <span
                  className="text-secondary"
                >
                   / 
                </span>
                <span
                  className="text-danger"
                >
                  -
                  12
                </span>
              </div>
            </div>
            <div />
          </div>
          <div
            className="chat__body"
          >
            <div
              className="chat__message chat__message_reversed"
              key="0"
            >
              <div
                className="chat__message__avatar"
              >
                <Avatar
                  url="url"
                >
                  <img
                    alt="Avatar"
                    className="avatar"
                    src="url"
                  />
                </Avatar>
              </div>
              <div>
                <div
                  className="chat__message__text chat__message__text_incoming"
                />
                <div
                  className="chat__message__date chat__message__date_align-right"
                >
                  9/1/2020, 00:00:00
                </div>
              </div>
            </div>
            <div
              className="chat__message"
              key="1"
            >
              <div
                className="chat__message__avatar"
              >
                <Avatar
                  url="https://s3.amazonaws.com/uifaces/faces/twitter/ssbb_me/128.jpg"
                >
                  <img
                    alt="Avatar"
                    className="avatar"
                    src="https://s3.amazonaws.com/uifaces/faces/twitter/ssbb_me/128.jpg"
                  />
                </Avatar>
              </div>
              <div>
                <div
                  className="chat__message__text chat__message__text_outgoing"
                />
                <div
                  className="chat__message__date"
                >
                  9/1/2020, 00:00:00
                </div>
              </div>
            </div>
            <div />
          </div>
          <Formik
            initialValues={
              Object {
                "message": "",
              }
            }
            onSubmit={[Function]}
            validate={[Function]}
          >
            <Form>
              <form
                action="#"
                onReset={[Function]}
                onSubmit={[Function]}
              >
                <div
                  className="chat__footer"
                >
                  <div
                    className="chat__footer__left-area"
                  >
                    <Field
                      name="message"
                      placeholder="Type your message..."
                    >
                      <input
                        name="message"
                        onBlur={[Function]}
                        onChange={[Function]}
                        placeholder="Type your message..."
                        value=""
                      />
                    </Field>
                  </div>
                  <div
                    className="chat__footer__right-area"
                  >
                    <button
                      type="submit"
                    >
                      Send
                    </button>
                  </div>
                </div>
              </form>
            </Form>
          </Formik>
        </div>
      </Memo(Chat)>
    </Route>
  </Router>
</Provider>
`;
